worldedit-core/src/main/java/com/sk89q/worldedit/extent/reorder/MultiStageReorder.java

```diff
@@ -94,58 +99,76 @@ public void setEnabled(boolean enabled) {
-
         if (!enabled) {
             return super.setBlock(location, block);
         }
 
-        if (Blocks.shouldPlaceLast(block.getBlockType())) {
-            // Place torches, etc. last
-            stage2.add(location, block);
-            return !existing.equalsFuzzy(block);
-        } else if (Blocks.shouldPlaceFinal(block.getBlockType())) {
-            // Place signs, reed, etc even later
-            stage3.add(location, block);
-            return !existing.equalsFuzzy(block);
-        } else if (Blocks.shouldPlaceLast(existing.getBlockType())) {
+        BlockState existing = getBlock(location);
+        int priority = getPlacementPriority(block);
+        int srcPriority = getPlacementPriority(existing);
+
+        if (srcPriority == 1 || srcPriority == 2) {
             // Destroy torches, etc. first
             super.setBlock(location, BlockTypes.AIR.getDefaultState());
             return super.setBlock(location, block);
-        } else {
-            stage1.add(location, block);
-            return !existing.equalsFuzzy(block);
         }
+
+        stages.get(priority).add(location, block);
+        return !existing.equalsFuzzy(block);
     }
 
     @Override
     public Operation commitBefore() {
-        return new OperationQueue(
-                new SetLocatedBlocks(
-                        getExtent(),
-                        Iterables.concat(stage1, stage2)),
-                new Stage3Committer());
+        List<Operation> operations = new ArrayList<>();
+        for (int i = 0; i < stages.size() - 1; ++i) {
+            operations.add(new SetLocatedBlocks(getExtent(), stages.get(i)));
+        }
+
+        operations.add(new FinalStageCommitter());
+        return new OperationQueue(operations);
     }
 
-    private class Stage3Committer implements Operation {
+    private class FinalStageCommitter implements Operation {
+        private Extent extent = getExtent();
 
-        @Override
-        public Operation resume(RunContext run) throws WorldEditException {
-            Extent extent = getExtent();
+        private final Set<BlockVector3> blocks = new HashSet<>();
+        private final Map<BlockVector3, BlockStateHolder> blockTypes = new HashMap<>();
 
-            final Set<BlockVector3> blocks = new HashSet<>();
-            final Map<BlockVector3, BlockStateHolder> blockTypes = new HashMap<>();
-            for (LocatedBlock entry : stage3) {
+        public FinalStageCommitter() {
+            for (LocatedBlock entry : stages.get(stages.size() - 1)) {
                 final BlockVector3 pt = entry.getLocation();
                 blocks.add(pt);
                 blockTypes.put(pt, entry.getBlock());
             }
+        }
 
+        @Override
+        public Operation resume(RunContext run) throws WorldEditException {
             while (!blocks.isEmpty()) {
                 BlockVector3 current = blocks.iterator().next();
                 if (!blocks.contains(current)) {
```

Comment:
Can this ever be false? You create a new iterator, get the first element, and check if it's still in the same collection.

Reply:
Additionally, this isn't very useful, because if this is `false`, nothing will change -- `blocks` doesn't get modified.